editor/src/viewport_tools/tools/freehand.rs

```diff
@@ -0,0 +1,192 @@
+use crate::document::DocumentMessageHandler;
+use crate::frontend::utility_types::MouseCursorIcon;
+use crate::input::keyboard::MouseMotion;
+use crate::input::InputPreprocessorMessageHandler;
+use crate::message_prelude::*;
+use crate::misc::{HintData, HintGroup, HintInfo};
+use crate::viewport_tools::tool::{DocumentToolData, Fsm, ToolActionHandlerData, ToolType};
+use crate::viewport_tools::tool_options::ToolOptions;
+
+use graphene::layers::style;
+use graphene::Operation;
+
+use glam::{DAffine2, DVec2};
+use serde::{Deserialize, Serialize};
+
+#[derive(Default)]
+pub struct Freehand {
+	fsm_state: FreehandToolFsmState,
+	data: FreehandToolData,
+}
+
+#[remain::sorted]
+#[impl_message(Message, ToolMessage, Freehand)]
+#[derive(PartialEq, Clone, Debug, Hash, Serialize, Deserialize)]
+pub enum FreehandMessage {
+	// Standard messages
+	#[remain::unsorted]
+	Abort,
+
+	// Tool-specific messages
+	DragStart,
+	DragStop,
+	PointerMove,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum FreehandToolFsmState {
+	Ready,
+	Drawing,
+}
+
+impl<'a> MessageHandler<ToolMessage, ToolActionHandlerData<'a>> for Freehand {
+	fn process_action(&mut self, action: ToolMessage, data: ToolActionHandlerData<'a>, responses: &mut VecDeque<Message>) {
+		if action == ToolMessage::UpdateHints {
+			self.fsm_state.update_hints(responses);
+			return;
+		}
+
+		if action == ToolMessage::UpdateCursor {
+			self.fsm_state.update_cursor(responses);
+			return;
+		}
+
+		let new_state = self.fsm_state.transition(action, data.0, data.1, &mut self.data, data.2, responses);
+
+		if self.fsm_state != new_state {
+			self.fsm_state = new_state;
+			self.fsm_state.update_hints(responses);
+			self.fsm_state.update_cursor(responses);
+		}
+	}
+
+	fn actions(&self) -> ActionList {
+		use FreehandToolFsmState::*;
+
+		match self.fsm_state {
+			Ready => actions!(FreehandMessageDiscriminant; DragStart, DragStop, Abort),
+			Drawing => actions!(FreehandMessageDiscriminant; DragStop, PointerMove, Abort),
+		}
+	}
+}
+
+impl Default for FreehandToolFsmState {
+	fn default() -> Self {
+		FreehandToolFsmState::Ready
+	}
+}
+#[derive(Clone, Debug, Default)]
+struct FreehandToolData {
+	points: Vec<DVec2>,
+	weight: u32,
+	path: Option<Vec<LayerId>>,
+	layer_exists: bool,
```

Comment:
Shouldn't this always be the case when path is some? You appear not to be using this in your code anyway.

Reply:
Removed this from pen tool as well. Was also unused there.